File: sdk/csharp/HuwutongSdk/HwtClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace HuwutongSdk
{
    /// <summary>
    /// HWT License Client — C# SDK (M2-21)
    ///
    /// 基于统一错误码标准 M2-34
    /// 支持: activate / validate / deactivate / checkFeature / getOfflineLicense / verifyOffline
    /// </summary>
    public class HwtClient
    {
        private readonly string _apiKey;
        private readonly string _host;
        private readonly string _secretKey;
        private readonly HttpClient _http;
        private readonly JsonSerializerOptions _jsonOptions;

        /// <summary>
        /// 创建 HWT 客户端
        /// </summary>
        /// <param name="apiKey">API Key</param>
        /// <param name="host">API 地址，默认 https://api.huwutong.com</param>
        /// <param name="secretKey">HMAC 签名密钥（可选）</param>
        /// <param name="timeoutSeconds">超时秒数，默认 30</param>
        public HwtClient(string apiKey, string host = "https://api.huwutong.com", string? secretKey = null, int timeoutSeconds = 30)
        {
            _apiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));
            _host = host?.TrimEnd('/') ?? throw new ArgumentNullException(nameof(host));
            _secretKey = secretKey ?? "";
            _http = new HttpClient { Timeout = TimeSpan.FromSeconds(timeoutSeconds) };
            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            };
        }

        #region Core API

        /// <summary>激活 License</summary>
        public async Task<ActivationResult> Activate(string licenseKey, Dictionary<string, string> deviceInfo)
        {
            var payload = new Dictionary<string, object>
            {
                ["license_key"] = licenseKey,
                ["device_info"] = deviceInfo,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            return await PostAsync<ActivationResult>("/api/license/activate", payload);
        }

        /// <summary>验证 License</summary>
        public async Task<ValidationResult> Validate(string licenseKey, Dictionary<string, string> deviceInfo)
        {
            var payload = new Dictionary<string, object>
            {
                ["license_key"] = licenseKey,
                ["device_info"] = deviceInfo,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            return await PostAsync<ValidationResult>("/api/license/validate", payload);
        }

        /// <summary>停用 License</summary>
        public async Task<HwtResponse<object>> Deactivate(string licenseKey, Dictionary<string, string> deviceInfo)
        {
            var payload = new Dictionary<string, object>
            {
                ["license_key"] = licenseKey,
                ["device_info"] = deviceInfo,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            return await PostAsync<HwtResponse<object>>("/api/license/deactivate", payload);
        }

        /// <summary>检查 Feature Flag</summary>
        public async Task<HwtResponse<FeatureCheckResult>> CheckFeature(string licenseKey, string featureKey)
        {
            var payload = new Dictionary<string, object>
            {
                ["license_key"] = licenseKey,
                ["feature"] = featureKey,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            return await PostAsync<HwtResponse<FeatureCheckResult>>("/api/license/check-feature", payload);
        }

        /// <summary>获取离线 License 文件</summary>
        public async Task<HwtResponse<OfflineLicenseData>> GetOfflineLicense(string licenseKey)
        {
            var payload = new Dictionary<string, object>
            {
                ["license_key"] = licenseKey,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            return await PostAsync<HwtResponse<OfflineLicenseData>>("/api/offline/generate", payload);
        }

        /// <summary>验证离线 License</summary>
        public async Task<HwtResponse<VerifyOfflineData>> VerifyOffline(string licenseData)
        {
            var payload = new Dictionary<string, object>
            {
                ["license_data"] = licenseData,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            };
            return await PostAsync<HwtResponse<VerifyOfflineData>>("/api/offline/verify", payload);
        }

        #endregion

        #region Private

        private async Task<T> PostAsync<T>(string path, object payload)
        {
            var url = $"{_host}{path}";
            var json = JsonSerializer.Serialize(payload, _jsonOptions);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var request = new HttpRequestMessage(HttpMethod.Post, url)
            {
                Content = content
            };
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            if (!string.IsNullOrEmpty(_secretKey))
            {
                var nonce = Guid.NewGuid().ToString("N");
                var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
                var signature = ComputeHmac(json, nonce, timestamp);
                request.Headers.Add("X-Nonce", nonce);
                request.Headers.Add("X-Timestamp", timestamp);
                request.Headers.Add("X-Signature", signature);
            }

            try
            {
                var resp = await _http.SendAsync(request);
                var body = await resp.Content.ReadAsStringAsync();
                var result = JsonSerializer.Deserialize<ApiEnvelope<T>>(body, _jsonOptions);

                if (result?.Error != null)
                {
                    throw new HwtApiException(
                        result.Error.Code ?? "UNKNOWN",
                        result.Error.Message ?? "未知错误",
                        (int)resp.StatusCode
                    );
                }

                return result!.Data ?? throw new HwtApiException("NULL_DATA", "响应数据为空", (int)resp.StatusCode);
            }
            catch (HwtApiException) { throw; }
            catch (TaskCanceledException ex)
            {
                throw new HwtNetworkException("请求超时", ex);
            }
            catch (Exception ex)
            {
                throw new HwtNetworkException(ex.Message, ex);
            }
        }

        private string ComputeHmac(string json, string nonce, string timestamp)
        {
            var data = $"{json}{nonce}{timestamp}";
            var keyBytes = Encoding.UTF8.GetBytes(_secretKey);
            var dataBytes = Encoding.UTF8.GetBytes(data);
            var hash = HMACSHA256.HashData(keyBytes, dataBytes);
            return Convert.ToHexString(hash).ToLower();
        }

        #endregion
    }

    // 内部 API 信封
    internal class ApiEnvelope<T>
    {
        public bool Success { get; set; }
        public T? Data { get; set; }
        public ApiError? Error { get; set; }
    }

    public class FeatureCheckResult
    {
        public bool Enabled { get; set; }
        public string? Feature { get; set; }
    }

    public class OfflineLicenseData
    {
        public string? LicenseData { get; set; }
        public string? ExpiresAt { get; set; }
    }

    public class VerifyOfflineData
    {
        public bool IsValid { get; set; }
        public string? LicenseKey { get; set; }
        public string? Status { get; set; }
    }
}

File: sdk/flutter/lib/src/hwt_client.dart
import 'dart:convert';
import 'dart:math';
import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;
import 'hwt_exceptions.dart';
import 'models.dart';

/// HWT License Client — Flutter/Dart SDK (M2-21)
///
/// 基于统一错误码标准 M2-34
class HwtClient {
  final String apiKey;
  final String host;
  final String? secretKey;
  final Duration timeout;

  HwtClient({
    required this.apiKey,
    this.host = 'https://api.huwutong.com',
    this.secretKey,
    this.timeout = const Duration(seconds: 30),
  });

  // ── Public API ──

  /// 激活 License
  Future<ActivationResult> activate(String licenseKey, Map<String, String> deviceInfo) async {
    final payload = {
      'license_key': licenseKey,
      'device_info': deviceInfo,
      'timestamp': (DateTime.now().millisecondsSinceEpoch / 1000).round(),
    };
    return _post<ActivationResult>(
      '/api/license/activate',
      payload,
      (json) => ActivationResult.fromJson(json),
    );
  }

  /// 验证 License
  Future<ValidationResult> validate(String licenseKey, Map<String, String> deviceInfo) async {
    final payload = {
      'license_key': licenseKey,
      'device_info': deviceInfo,
      'timestamp': (DateTime.now().millisecondsSinceEpoch / 1000).round(),
    };
    return _post<ValidationResult>(
      '/api/license/validate',
      payload,
      (json) => ValidationResult.fromJson(json),
    );
  }

  /// 停用 License
  Future<Map<String, dynamic>> deactivate(String licenseKey, Map<String, String> deviceInfo) async {
    final payload = {
      'license_key': licenseKey,
      'device_info': deviceInfo,
      'timestamp': (DateTime.now().millisecondsSinceEpoch / 1000).round(),
    };
    return _post<Map<String, dynamic>>(
      '/api/license/deactivate',
      payload,
      (json) => json,
    );
  }

  /// 检查 Feature Flag
  Future<Map<String, dynamic>> checkFeature(String licenseKey, String featureKey) async {
    final payload = {
      'license_key': licenseKey,
      'feature': featureKey,
      'timestamp': (DateTime.now().millisecondsSinceEpoch / 1000).round(),
    };
    return _post<Map<String, dynamic>>(
      '/api/license/check-feature',
      payload,
      (json) => json,
    );
  }

  /// 获取离线 License
  Future<Map<String, dynamic>> getOfflineLicense(String licenseKey) async {
    final payload = {
      'license_key': licenseKey,
      'timestamp': (DateTime.now().millisecondsSinceEpoch / 1000).round(),
    };
    return _post<Map<String, dynamic>>(
      '/api/offline/generate',
      payload,
      (json) => json,
    );
  }

  /// 验证离线 License
  Future<Map<String, dynamic>> verifyOffline(String licenseData) async {
    final payload = {
      'license_data': licenseData,
      'timestamp': (DateTime.now().millisecondsSinceEpoch / 1000).round(),
    };
    return _post<Map<String, dynamic>>(
      '/api/offline/verify',
      payload,
      (json) => json,
    );
  }

  // ── Internal ──

  Future<T> _post<T>(
    String path,
    Map<String, dynamic> payload,
    T Function(Map<String, dynamic> json) fromJson,
  ) async {
    final url = Uri.parse('$host$path');
    final body = jsonEncode(payload);

    final headers = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'Authorization': 'Bearer $apiKey',
    };

    if (secretKey != null && secretKey!.isNotEmpty) {
      final nonce = _generateNonce();
      final timestamp = (DateTime.now().millisecondsSinceEpoch / 1000).round().toString();
      final signature = _hmac(body, nonce, timestamp);
      headers['X-Nonce'] = nonce;
      headers['X-Timestamp'] = timestamp;
      headers['X-Signature'] = signature;
    }

    try {
      final response = await http
          .post(url, headers: headers, body: body)
          .timeout(timeout);

      final result = jsonDecode(response.body) as Map<String, dynamic>;

      if (result.containsKey('error') && result['error'] != null) {
        final error = ApiError.fromJson(result['error'] as Map<String, dynamic>);
        throw HwtApiException(
          error.code ?? 'UNKNOWN',
          error.message ?? 'Unknown error',
        );
      }

      return fromJson(result);
    } on HwtApiException {
      rethrow;
    } on http.ClientException catch (e) {
      throw HwtNetworkException(e.message);
    } on Exception catch (e) {
      throw HwtNetworkException(e.toString());
    }
  }

  String _generateNonce() {
    final rand = Random.secure();
    final bytes = List<int>.generate(16, (_) => rand.nextInt(256));
    return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }

  String _hmac(String body, String nonce, String timestamp) {
    final data = '$body$nonce$timestamp';
    final hmacSha256 = Hmac(sha256, utf8.encode(secretKey!));
    final digest = hmacSha256.convert(utf8.encode(data));
    return digest.toString();
  }
}
